src/scheduler/MpiWorld.cpp

```diff
@@ -714,17 +787,65 @@ void MpiWorld::awaitAsyncRequest(int requestId)
 {
     SPDLOG_TRACE("MPI - await {}", requestId);
 
-    auto it = futureMap.find(requestId);
-    if (it == futureMap.end()) {
-        throw std::runtime_error(
-          fmt::format("Error: waiting for unrecognized request {}", requestId));
+    // Get the corresponding send and recv ranks
+    auto it = reqIdToRanks.find(requestId);
+    if (it == reqIdToRanks.end()) {
+        SPDLOG_ERROR("Asynchronous request id not recognized: {}", requestId);
+        throw std::runtime_error("Unrecognized async request id");
+    }
+    int sendRank = it->second.first;
+    int recvRank = it->second.second;
+    reqIdToRanks.erase(it);
+
+    // If awaiting an isend request, return immediately as our transport layer
+    // is asynchronous, thus request is already handled
+    if (sendRank == MPI_IS_ISEND_REQUEST && recvRank == MPI_IS_ISEND_REQUEST) {
+        return;
     }
 
-    // This call blocks until requestId has finished.
-    it->second.wait();
-    futureMap.erase(it);
+    std::shared_ptr<faabric::scheduler::UnackedMessageBuffer> umb =
+      getUnackedMessageBuffer(sendRank, recvRank);
+
+    // The request id must be in the UMB, as an irecv must happen before an
+    // await. It is very likely that if the assert were to fail, the previous
+    // check would have also failed, thus why we assert here.
```

Comment:
The assertions in this method and several others in this PR seem to be guarding against something that might fail at runtime, e.g. a message not getting delivered due to some networking issue or an issue with the running application logic sending and receiving from the wrong ranks. 

This comment refers to "a previous check" but it's hard to tell what that is, and what error condition it's actually describing. Perhaps something like "If the message we are awaiting isn't present in the unacked message buffer, then it either hasn't been sent, or hasn't been delivered". If it hasn't been sent then this could be an issue with the MPI application we're executing, or it could be a network failure, or it could be the logic of our implementation. The first two cases warrant an exception with a meaninful error message.